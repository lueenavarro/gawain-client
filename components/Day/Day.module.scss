@use "styles/colors";

.day {
  border: 1px solid colors.$border;
  border-left-width: 0.5px;
  border-right-width: 0.5px;
  max-width: 500px;
  height: 100%;
  padding: 5px 0;
  display: flex;
  flex-direction: column;

  &__name {
    text-align: center;
    margin-bottom: 5px;
  }
  &__date {
    text-align: center;
    margin-top: 0;
    font-weight: normal;
  }

  &__bg {
    background-image: linear-gradient(#eee 0.1em, transparent 0.1em);
    background-size: 100% 30px;
    flex: 1;
  }
}
