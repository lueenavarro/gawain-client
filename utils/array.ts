import { concat, slice } from "lodash";

export const insert = (array: Array<any>, index: number, item: any) => {
  if (index < 0) {
    throw new Error("Invalid index");
  }

  return concat(slice(array, 0, index), item, slice(array, index));
};

export const getMiddle = (array: Array<any>) => slice(array, 1, -1);
