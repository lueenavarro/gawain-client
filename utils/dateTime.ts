import moment from "moment";

export const formatDate = (date: string | Date, format = "MMMM D, YYYY") =>
  date ? moment(date).format(format) : date;

export const formatDateTime = (
  date: string | Date,
  format = "MMMM D, YYYY, h:mm A"
) => (date ? moment(date).format(format) : date);

export const formatDateForApi = (date: string | Date, format = "YYYY-MM-DD") =>
  date ? moment(date).format(format) : date;

export const addDays = (date: Date, days: number) => {
  const result = new Date(date.getTime());
  result.setDate(result.getDate() + days);
  return result;
};
