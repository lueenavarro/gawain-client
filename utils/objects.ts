import { cloneDeep } from "lodash";

export const sliceObjects = (object, start?: number, end?: number) => {
    const clone = cloneDeep(object);
    Object.keys(clone)
      .slice(start, end)
      .forEach((key) => delete clone[key]);
    return clone;
  };