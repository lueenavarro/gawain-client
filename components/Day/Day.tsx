import React from "react";
import PropTypes from "prop-types";

import AddTask from "components/AddTask";
import { Task, TaskClone } from "components/Task";
import IDraggable from "shared/components/Draggable";
import IDroppable from "shared/components/Droppable";
import { formatDate } from "utils/dateTime";

import styles from "./Day.module.scss";

const Day = ({ taskList, onAddTask, onRemove, onComplete }) => (
  <div className={styles.day}>
    <h3 className={styles["day__name"]} data-testid="day">
      {taskList.day}
    </h3>
    <h5 className={styles["day__date"]} data-testid="date">
      {formatDate(taskList.date)}
    </h5>
    <div className={styles["day__bg"]}>
      <IDroppable
        droppableId={taskList.date}
        list={taskList.tasks}
        cloneParent={TaskClone}
      >
        {taskList.tasks.map((task, index) => (
          <IDraggable key={task._id} draggableId={task._id} index={index}>
            <Task
              task={task}
              onRemove={(...args) => onRemove(...args, taskList.date)}
              onComplete={(...args) => onComplete(...args, taskList.date)}
            />
          </IDraggable>
        ))}
        <AddTask onAddTask={(...args) => onAddTask(...args, taskList.date)} />
      </IDroppable>
    </div>
  </div>
);

Day.propTypes = {
  taskList: PropTypes.any.isRequired,
  onAddTask: PropTypes.func.isRequired,
  onRemove: PropTypes.func.isRequired,
};

export default Day;
