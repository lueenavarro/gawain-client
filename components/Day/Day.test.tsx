import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";

import { render } from "test/react";
import { formatDate } from "utils/dateTime";
import Day from "./Day";

jest.mock("shared/components/Droppable", () => ({
  __esModule: true,
  default: ({ children }) => <div>{children}</div>,
}));

jest.mock("shared/components/Draggable", () => ({
  __esModule: true,
  default: ({ children, index }) => <div key={index}>{children}</div>,
}));

jest.mock("components/Task", () => ({
  __esModule: true,
  default: ({ task }) => <span data-testid="task">{task.task}</span>,
}));

jest.mock("components/AddTask", () => ({
  __esModule: true,
  default: () => <input type="text" />,
}));

jest.mock("utils/dateTime", () => ({
  formatDate: jest.fn((_date) => {}),
}));

const fakeTasks = [
  {
    _id: "1",
    task: "Go To Market",
    onRemove: jest.fn(() => {}),
  },
  {
    _id: "2",
    task: "Go To Gym",
    onRemove: jest.fn(() => {}),
  },
];

const mockOnAddTask = jest.fn(() => {});
const mockOnRemove = jest.fn(() => {});
const mockOnComplete = jest.fn(() => {});

test("render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Day
      day="Monday"
      date="2020-02-28"
      tasks={[]}
      onAddTask={mockOnAddTask}
      onRemove={mockOnRemove}
      onComplete={mockOnComplete}
    />,
    div
  );
});

test("render day, date and tasks correctly", () => {
  (formatDate as jest.Mock).mockReturnValue("February 28");

  const { getByTestId, getAllByTestId } = render(
    <Day
      day="Monday"
      date="2020-02-28"
      tasks={fakeTasks}
      onAddTask={mockOnAddTask}
      onRemove={mockOnRemove}
      onComplete={mockOnComplete}
    />
  );

  expect(getByTestId("day").innerHTML).toBe("Monday");
  expect(getByTestId("date").innerHTML).toBe("February 28");
  expect(getAllByTestId("task").length).toBe(2);
  expect(getAllByTestId("task")[0].innerHTML).toBe("Go To Market");
  expect(getAllByTestId("task")[1].innerHTML).toBe("Go To Gym");
});

test("snapshot matches", () => {
  const tree = renderer
    .create(
      <Day
        day="Monday"
        date="2020-02-28"
        tasks={fakeTasks}
        onAddTask={mockOnAddTask}
        onRemove={mockOnRemove}
        onComplete={mockOnComplete}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
