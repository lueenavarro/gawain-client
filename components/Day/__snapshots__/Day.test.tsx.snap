// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
<div
  className="day"
>
  <h3
    className="day__name"
    data-testid="day"
  >
    Monday
  </h3>
  <h5
    className="day__date"
    data-testid="date"
  >
    February 28
  </h5>
  <div
    className="day__bg"
  >
    <div>
      <div>
        <span
          data-testid="task"
        >
          Go To Market
        </span>
      </div>
      <div>
        <span
          data-testid="task"
        >
          Go To Gym
        </span>
      </div>
      <input
        type="text"
      />
    </div>
  </div>
</div>
`;
